test/params_api_test.c

```diff
@@ -401,6 +401,34 @@ static int test_param_bignum(int n)
     return ret;
 }
 
+static int test_param_bignum_pad(void)
+{
+    unsigned char buf[MAX_LEN];
+    OSSL_PARAM param;
+    BIGNUM *bn, *out = NULL;
+    size_t i;
+    int ret = 0;
+
+    param = OSSL_PARAM_construct_BN("bn", buf, sizeof(buf));
+    if (!TEST_ptr(bn = BN_new())
+            || !TEST_ptr(out = BN_new())
+            || !TEST_true(BN_set_word(bn, 54321)))
+        goto err;
+    for (i = sizeof(BN_ULONG); i < MAX_LEN; i++) {
+        if (!TEST_true(OSSL_PARAM_set_BN_pad(&param, bn, i))
+                || !TEST_size_t_eq(param.return_size, i)
+                || !TEST_true(OSSL_PARAM_get_BN(&param, &out))
```

Comment:
I see where your change of `OSSL_PARAM_get_BN` comes from.  It it were me, I would just `!TEST_true(BN_native2bn(buf, param->return_size, out))`

Reply:
This complete circumvents the problem......

The param owner is permitted to call `OSSL_PARAM_set_BN` with an undersized BN and then pass it over.

Reply:
`OSSL_PARAM_set_BN_pad()` is designed for the "receiving end", just as all the other `set` and `get` functions, is it not?

The `OSSL_PARAM` owner calls `OSSL_PARAM_construct_BN()` with the appropriate buffer size.

See, I think you're confusing yourself with this test, where you're clearly mixing the "sending end" and the "receiving end"...  which is fine, *if you're clear about which call belongs in what end*.  I would suggest dividing up this test in two parts to make things clearer for yourself.

Reply:
I view `OSSL_PARAM_set_BN_pad()` as being applicable on **both** the sending and receiving sides.  Unlike all the other types, we cannot create an OSSL_PARAM containing a BN without having extra storage.  The _to native_ conversion mandates a memory copy, this isn't true for the string types or the basic numeric types -- these can be their own storage.

I think it quite reasonable for the sender to use `OSSL_PARAM_construct_BN()` to create a PARAM and `OSSL_PARAM_set_BN_pad()` to populate it.

I'm beginning to think that a BN PARAM type might avoid a lot of the gymnastics we're being forced through.

Reply:
> I'm beginning to think that a BN PARAM type might avoid a lot of the gymnastics we're being forced through.

How?  Are you going to just point directly to the BIGNUM, i.e. force everyone to use that specific storage format for big numbers?  In that case, *we need to have the format very well documented*, and the BIGNUM type will have to continue to be an open and unchanging structure for the foreseeable future.  That, or lock every in on our BIGNUM implementation.

Whichever way we turn with BIGNUMs, there are consequences to consider.

Reply:
Of course, "lock in" is relative.  When we made the nCipher (CHIL) engine, we had to deal with exactly that, the result being a couple of conversion routines for juggling big numbers back and forth between OpenSSL's BN and nCipher's MPI.  Look for `mpi2bn` and `bn2mpi` in https://github.com/engine-corner/engine-chil/blob/master/e_chil.c to see

Reply:
However, back to `OSSL_PARAM_set_BN()` / `OSSL_PARAM_set_BN_pad()` for the "sending" end, I'll suggest again to make it a separate function.  I've said this what feels like a gazillion times already.  That's the only way I can see avoiding dicy corners.

Reply:
I was thinking of exposing the BN_ULONG array as the passing mechanism.  Plus the length of the array.  We don't need the sign.  The flags might be a loss.

I'm tempted to ditch this PR as too difficult.